compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -3155,7 +3164,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 (root, pred) if seen_preds.contains(&pred) || seen_preds.contains(&root) => {}
                 _ => continue,
             }
-            error.obligation.cause.span = span;
+            error.obligation.cause.span = idx_span;
+
+            // If the index value is a double borrow, that can cause typeck errors
+            // that can be easily resolved by removing the borrow from the expression.
+            // We check for that here and provide a suggestion in a custom obligation
+            // cause code.
+            if let hir::ExprKind::AddrOf(_, _, idx) = idx.kind {
+                let idx_t = self.typeck_results.borrow().expr_ty(idx);
+                let mut autoderef = self.autoderef(base.span, base_t);
+                let mut result = None;
+                while result.is_none() && autoderef.next().is_some() {
+                    result = self.try_index_step(expr, base, &autoderef, idx_t, idx);
```

Comment:
`try_index_step` still registers things into the fnctxt's fulfullment context :(

I think it's simpler to be creating your own `Index` trait obligation here -- the only thing that you'd be losing is the `[u8; N]` -> `[u8]` deref step, which I guess you could replicate here.